### PR TITLE
Fix From As in Dockerfile

### PR DESCRIPTION
**Pull request by TomSweeneyRedHat**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

This is the second stab at this.  This time without changes to imagebuilder.  I need to test more and add tests, but would appreciate reviews.  The hang up was we weren't parsing out the "AS image" part of the "FROM image AS image" command in a Dockerfile.  It was not being noted in the output and worse yet Buildah didn't know what to do with it.  Also each "FROM" in the Dockerfile causes imagebuilder to create a stage for each one.  The environment variables created in the first stage are not sent to the second or later stages.  This change creates a slice to collect and use them all.  I'm not sure if other things should or should not also be collected, happy to hear opinions.  I initially started by collecting everything in the Docker Config struct, but eventually narrowed that down just to the Env slice in that struct.  I could revert if necessary.

Test run with Conformance test Dockerfile that uncovered this issue.
```
# buildah bud -t tom -f ~/Dockerfile .
STEP 1: FROM centos:7 AS base
STEP 2: RUN touch /1
STEP 3: ENV LOCAL=/1
STEP 4: RUN find $LOCAL
/1
STEP 5: FROM base
STEP 6: RUN find $LOCAL 
/1
STEP 7: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/tom:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:1d31b5806ba40b5f67bde96f18a181668348934a44c9253b420d5f04cfb4e37a
Copying blob sha256:021b1264e8efc3ed17a3ec27a43d7ee8b5cd1e437860341c64f34625ab005aca
 167 B / 167 B [============================================================] 0s
Copying config sha256:30ce4fa0ea522c7115a9e3a2104bb0e12fcf01469a36fd32a167f2c6019b9521
 1.17 KiB / 1.17 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
--> 30ce4fa0ea522c7115a9e3a2104bb0e12fcf01469a36fd32a167f2c6019b9521

# ctr=$(buildah from tom)

# buildah run $ctr /bin/bash
[root@mrsdalloway /]# echo $LOCAL
/1

```

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
@nalind something bizzare is going on with this PR/review, hold off for now.  Will reping.

**Comment by TomSweeneyRedHat** [comment]
@nalind @umohnani8 @rhatdan PTAL

**Comment by rhatdan** [comment]
@TomSweeneyRedHat If this is ready to merge can you remove WIP?

**Comment by rhatdan** [comment]
@giuseppe @vrothberg PTAL

**Comment by TomSweeneyRedHat** [comment]
Not yet ready to merge.  One of the bud tests is blowing up.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan ack, will do when I chunk in whatever changes necessary to get the test happy and also address the review comments.

**Comment by TomSweeneyRedHat** [comment]
Looks like a build flake on one test, going to rekick.  Testing went well last night, but need to run a few more this morn before removing the WIP.  Also need to add a test or two.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please.

**Comment by rhatdan** [comment]
Your PR Still says WIP, we can not merge it and will need to go through the test cycle again.

**Comment by TomSweeneyRedHat** [comment]
Yep, as I went to bed last night I'd  a "Oh NO!" thought, turned out to be OK after all.    Writing a bats test to test the new functionality, will remove WIP with that push.

File: imagebuildah/build.go [hunk 1]
```diff
@@ -1272,6 +1313,18 @@ func BuildDockerfiles(ctx context.Context, store storage.Store, options BuildOpt
 	return exec.Build(ctx, stages)
 }
 
+// updateImageConfig in the case of multiple stages,
+// add all the environment variables together.
+func updateImageConfig(config *docker.Config) {
+	if len(imageEnv) == 0 {
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: I thought I'd  tried that at one point and it failed with an accvio, but it seems to be working now.  I must of had a syntax issue prior I guess.  Changed and thaks.

File: imagebuildah/build.go [hunk 2]
```diff
@@ -747,10 +776,19 @@ func (b *Executor) resolveNameToImageRef() (types.ImageReference, error) {
...
+	ib := stage.Builder
+	node := stage.Node
 	checkForLayers := true
 	children := node.Children
 	commitName := b.output
+	b.containerIDs = nil
+
+	// If we had envvar's from previous stages, add them to this one.
+	for _, envSpec := range imageEnv {
+		ib.Env = append(ib.Env, envSpec)
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: changed

**Comment by TomSweeneyRedHat** [comment]
Looks like errors talking to the registry, going to retest.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please.

**Comment by TomSweeneyRedHat** [comment]
bot is still ignoring me, going to close and reopen to retest.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Needs a rebase.  Lets get this merged by Monday so we can finally release 1.4

**Comment by TomSweeneyRedHat** [comment]
My other PR for the bud tests messed up this one.  Before merging I'd really like @nalind and @umohnani8 to PTAL

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please

**Comment by TomSweeneyRedHat** [comment]
@rhatdan, the bot's ignoring me.  Can you give it a kick for me?

**Comment by nalind** [comment]
bot, retest this please

**Comment by umohnani8** [comment]
Changes LGTM

**Comment by rhatdan** [comment]
bot, retest this please

File: imagebuildah/build.go [hunk 3]
```diff
@@ -646,10 +664,18 @@ func (b *Executor) Prepare(ctx context.Context, ib *imagebuilder.Builder, node *
...
 		Format:                b.outputFormat,
 	}
-	builder, err := buildah.NewBuilder(ctx, b.store, builderOptions)
-	if err != nil {
-		return errors.Wrapf(err, "error creating build container")
+
+	var builder *buildah.Builder
+	var err error
+	if asImageBuilderFound, ok := builderMap[from]; ok {
+		builder = asImageBuilderFound
```
Comment (nalind) [inline imagebuildah/build.go]: This needs to be changed to a fresh container based on an image that's based on the container.  One thing we see done in Dockerfiles is multiple stages referring to the same previous stage, as in https://gist.github.com/bparees/f1c8b734397481af465452bd60ad718e, and in those cases we want changes made in one stage to not affect what the other starts with.

Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: This is exactly why I wanted you to review this.  TYVM for the catch and I'll see what I can do to change the code to make this happen.

**Comment by TomSweeneyRedHat** [comment]
I've pushed the next commit that from what I can see addresses @nalind's concerns.  FIngers crossed on the test ci.

File: imagebuildah/build.go [hunk 4]
```diff
@@ -1200,6 +1244,12 @@ func (b *Executor) Build(ctx context.Context, stages imagebuilder.Stages) error
 			return errors.Errorf("Failed to cleanup intermediate containers")
 		}
 	}
+	// Remove intermediate images that we created for AS clause handling
+	for _, value := range imageMap {
+		if err := b.store.Delete(value); err != nil {
+			logrus.Debugf("unable to remove intermediate image %q: %v", value, err)
+		}
+	}
```
Comment (TomSweeneyRedHat) [inline imagebuildah/build.go]: @nalind I initially used b.store.DeleteImage(value, true) at line 1249, however it always failed into the Debugf statement and left the extra image behind.  I think it's safe to delete it outright like this, but wanted to verify with you.

**Comment by TomSweeneyRedHat** [comment]
Red Hat CI bombed with a registry error, going to retest.  @rhatdan if still not happy and/or the bot ignores me, can you retest in the a.m. please?

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
Tests pass, I want @umohnani8 and @nalind to pass this.

**Comment by umohnani8** [comment]
LGTM

File: imagebuildah/build.go (hunk 4 above)
Comment (nalind) [reply imagebuildah/build.go]: What was the error that `DeleteImage()` was returning?  Plain `Delete()` skips over some safety checks that we might want to not skip.

File: imagebuildah/build.go [hunk 5]
```diff
@@ -222,6 +222,8 @@ type Executor struct {
 	containerIDs                   []string // Stores the IDs of the successful intermediate containers used during layer build
 }
 
+var imageMap map[string]string
```
Comment (nalind) [inline imagebuildah/build.go]: Does this need to be global, rather than a member of the `Executor`?

Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: It's not in the executor, the curly at line 223 finishes up the executor.  However to make it stand out more, I'll move the declaration up to line 45ish, just after the const definition and before the Mount  definition.

File: imagebuildah/build.go (hunk 4 above)
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: "image is in use by a container" is the error.  For a test, I replaced the debug statement with a printf and prepended "TOM:" to it to make it stand out.  I'm not sure where the container/image is that it's complaining about, I don't see it in the output nor in the 'buildah images/containers output'.   I thought I had seen it yesterday though.


Test output:

```
# buildah bud -t tom -f ~/Dockerfile.fromas .
STEP 1: FROM centos:7 AS base
Getting image source signatures
Copying blob sha256:256b176beaff7815db2a93ee2071621ae88f451bb1e198ca73010ed5bba79b65
 71.23 MiB / 71.23 MiB [===================================================] 10s
Copying config sha256:5182e96772bf11f4b912658e265dfe0db8bd314475443b6434ea708784192892
 2.13 KiB / 2.13 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
STEP 2: RUN touch /1
STEP 3: ENV LOCAL=/1
STEP 4: RUN find $LOCAL
/1
STEP 5: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/tom:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:1d31b5806ba40b5f67bde96f18a181668348934a44c9253b420d5f04cfb4e37a
Copying blob sha256:a034a2927de80dc639e5e5f0eece366629321a3cbee05c2e49e24af6fe49443a
 167 B / 167 B [============================================================] 0s
Copying config sha256:821f92550dcdef1d20bc9f5cddc2fc2d8a04197592a5c062ed0ce9a8fdc3a88e
 1.18 KiB / 1.18 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
--> 821f92550dcdef1d20bc9f5cddc2fc2d8a04197592a5c062ed0ce9a8fdc3a88e
STEP 6: FROM base
STEP 7: RUN find $LOCAL
/1
STEP 8: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]localhost/tom:latest
Getting image source signatures
Skipping fetch of repeat blob sha256:1d31b5806ba40b5f67bde96f18a181668348934a44c9253b420d5f04cfb4e37a
Skipping fetch of repeat blob sha256:d2f84891591559baf3edea8db4273078145c71ba4d590d0a8368984a66afa42c
Copying blob sha256:4f4fb700ef54461cfa02571ae0db9a0dc1e0cdb5577484a6d75e68dc38e8acc1
 32 B / 32 B [==============================================================] 0s
Copying config sha256:46d9491b593562217acc42482afc9f74ffa57fc471d982eda43755ef1fe9db44
 1.29 KiB / 1.29 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
--> 46d9491b593562217acc42482afc9f74ffa57fc471d982eda43755ef1fe9db44
TOM: unable to remove intermediate image 821f92550dcdef1d20bc9f5cddc2fc2d8a04197592a5c062ed0ce9a8fdc3a88e: Image used by 3a07f584e4de06e1e9c7536caea12c52c1d270e86c19b0dcc17f3c2756ae3c13: image is in use by a container

```

File: imagebuildah/build.go (hunk 5 above)
Comment (nalind) [reply imagebuildah/build.go]: Sorry, I was being unclear:  why isn't it part of `Executor`?  If we had two builds going in a single process, things could get confused.

Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: To be honest, hadn't considered the multi build from one process scenario.   Will try moving it to the executor.

File: imagebuildah/build.go (hunk 4 above)
Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: For the purpose of this PR then, should I stick with DeleteImage or the straight Delete call?

Comment (nalind) [reply imagebuildah/build.go]: `DeleteImage` is heavily preferred - `Delete` removes the record of the image, but not its layers.

Comment (TomSweeneyRedHat) [reply imagebuildah/build.go]: That's what I thought, but thought I'd just double check as using it means we'll likely leave an image behind.

**Comment by TomSweeneyRedHat** [comment]
I've updated the PR based on Nalin's latest comments.  I'm firmly convinced the CI hates this PR, more network flakes, going to kick the bot again.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please.

**Comment by TomSweeneyRedHat** [comment]
blankeyt blank bot is ignoring me, going to close and reopen to force it's hand.

**Comment by TomSweeneyRedHat** [comment]
please behave ci

**Comment by TomSweeneyRedHat** [comment]
Another network flake, going to try kicking the bot again.

**Comment by TomSweeneyRedHat** [comment]
bot, retest this please

**Comment by TomSweeneyRedHat** [comment]
Looks like the bot ignored me, going to close and reopen.

**Comment by rhatdan** [comment]
Wait for @nalind Review and then merge this one.

**Comment by TomSweeneyRedHat** [comment]
I'm just happy to see green buttons on this finally....

File: tests/bud.bats [hunk 6]
```diff
@@ -897,3 +897,58 @@ load helpers
...
+
+  run buildah --debug=false images -a
+  [[ $output =~ "test1" ]]
+  [ "${status}" -eq 0 ]
+
+  ctr=$(buildah --debug=false from --signature-policy ${TESTSDIR}/policy.json test1)
+  run buildah --debug=false containers -a
+  [[ $output =~ "test1" ]]
+  [ "${status}" -eq 0 ]
+
```
Comment (nalind) [inline tests/bud.bats]: I think we need to check and ensure that $ctr/1 exists and $ctr/2 does not, somewhere around here, to verify that the third stage's starting point and the final image are based on the first named stage and not altered by anything we do for the second:
```
  run test -e $ctr/1
  [ "${status}" -eq 0 ]
  run test -e $ctr/2
  [ "${status}" -ne 0 ]
```
Either that or having the third stage perform the equivalent check as RUN statements.

Comment (nalind) [reply tests/bud.bats]: LGTM once we add this.

Comment (TomSweeneyRedHat) [reply tests/bud.bats]: good thought, will add that.

**Comment by TomSweeneyRedHat** [comment]
Tested @nalind suggested a test addition and it passed on my vm.  Have just push that and now I'm hoping for happy green buttons.

**Comment by nalind** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**